inference_demo.py

```diff
@@ -0,0 +1,186 @@
+        seq_len = input_ids.shape[1]
+
+        cache_position = torch.arange(seq_len, dtype=torch.int64, device=self.device)
+        generated_ids = torch.zeros(1, seq_len + max_new_tokens * 2, dtype=torch.int, device=self.device)
+        generated_ids[:, cache_position] = input_ids.to(self.device).to(torch.int)
+
+        past_key_values = StaticCache(self.model.config,
+                                      1,
+                                      seq_len + max_new_tokens * 2 + 1,
+                                      device=self.device,
+                                      dtype=torch.float16)
+
+        logits = self.model(
+            input_ids, cache_position=cache_position, past_key_values=past_key_values, return_dict=False, use_cache=True
+        )[0]
+        next_token = torch.argmax(logits[:, [-1]], dim=-1).to(torch.int)
+        generated_ids[:, [seq_len]] = next_token
+
+        torch._dynamo.config.capture_scalar_outputs = True
+
+        forward_time_s = []
+        with torch.no_grad():
+            # Compile the CUDA graph
+            if self.backend is None:
+                decode_one_tokens_compiled = decode_one_tokens
+            else:
+                decode_one_tokens_compiled = torch.compile(decode_one_tokens, mode='default', fullgraph=True)
+
+            # Generate tokens one by one
+            cache_position = torch.tensor([seq_len + 1], device='cuda')
+            for _ in range(1, max_new_tokens):
+                start_time = time.time()
+                next_token = decode_one_tokens_compiled(self.model, next_token.clone(), None, cache_position,
+                                                        past_key_values)
+                generated_ids[:, cache_position] = next_token.int()
+                end_time = time.time()
+                print(f'duration = {end_time - start_time}')
+                forward_time_s.append(end_time - start_time)
+
+                cache_position += 1
+
+        return self.tokenizer.decode(generated_ids[0]), forward_time_s
+
+
+if __name__ == "__main__":
+    import argparse
+
+    parser = argparse.ArgumentParser(add_help=True)
+
+    parser.add_argument(
+        "--pretrained_model_path",
+        type=str,
+        help="Path to the model to the pretrained model",
+    )
+    parser.add_argument(
+        "--compressed_model_path",
+        type=str,
+        help="Path to the compressed .pt model",
+    )
+    parser.add_argument(
+        "--execution_mode",
+        choices=[0, 1, 2],
+        required=True,
+        type=int,
+        help="If set to 0, will evaluate the dense pretrained model. "
+             "If set to 1, will evaluate the spqr-quantized model using HF"
+             "If set to 2, will evaluate the spqr-quantized model using torch .pt"
+    )
+
+    args = parser.parse_args()
+
+    m = Mode(args.execution_mode)
+
+    with torch.no_grad():
+        model = LLama(args.pretrained_model_path, args.compressed_model_path, m)
```

Comment:
what if model not LLama do we need to create , different class as an example above?

Reply:
It's just a thin wrapper class, I'll rename it to InferenceDemo, since it's only used within the context of the demo.